src/lambertw.jl

```diff
@@ -0,0 +1,386 @@
+import Base: convert
+#export lambertw, lambertwbp
+using Compat
+
+const euler =
+if isdefined(Base, :MathConstants)
+    Base.MathConstants.e
+else
+    e
+end
+
+const omega_const_bf_ = Ref{BigFloat}()
+
+function __init__()
+    omega_const_bf_[] =
+        parse(BigFloat,"0.5671432904097838729999686622103555497538157871865125081351310792230457930866845666932194")
+end
+
+#### Lambert W function ####
+
+const LAMBERTW_USE_NAN = false
+
+macro baddomain(v)
+    if LAMBERTW_USE_NAN
+        return :(return(NaN))
+    else
+        return esc(:(throw(DomainError($v))))
+    end
+end
+
+# Use Halley's root-finding method to find x = lambertw(z) with
+# initial point x.
+function _lambertw(z::T, x::T) where T <: Number
+    two_t = convert(T,2)
+    lastx = x
+    lastdiff = zero(T)
+    for i in 1:100
```

Comment:
Where does 100 come from? Seems like a rather arbitrary "magic number"; perhaps we could have that be configurable, e.g.
```julia
function _lambertw(z::T, x::T, maxiter::Int=100) where T<:Number
```

Reply:
Two things. 1) Some of this code was taken from mpmath. Maybe this number was, although I don't think a reason was given there  either. I should acknowledge the source somewhere, it has a MIT compatible license.  I will look at the python source. 2)  Yes, I think changing it is ok. How ? Maybe.... pick a number far larger than the number of iterations required to converge in most, or all cases.  But, make it small enough to finish the loop in a short time.  We could have it error or warn about lack of precision if the entire loop is completed. I'm pretty sure I had something like this in there at one point.